trinity/protocol/bcc/servers.py

```diff
@@ -0,0 +1,126 @@
+import functools
+import itertools
+
+from cytoolz import (
+    cons,
+    sliding_window,
+    take,
+)
+from eth_utils import (
+    to_tuple,
+)
+
+from cancel_token import CancelToken
+
+from p2p import protocol
+from p2p.peer import BasePeer
+from p2p.protocol import Command
+
+from eth.exceptions import BlockNotFound
+from eth.beacon.db.chain import BaseBeaconChainDB
+from eth.beacon.types.blocks import BaseBeaconBlock
+
+from trinity.protocol.common.servers import BaseRequestServer
+from trinity.protocol.bcc import commands
+from trinity.protocol.bcc.peer import (
+    BCCPeer,
+    BCCPeerPool,
+)
+
+from typing import (
+    cast,
+    Any,
+    Dict,
+    Iterable,
+    Set,
+    Type,
+)
+from eth_typing import (
+    Hash32,
+)
+
+
+class BCCRequestServer(BaseRequestServer):
+    subscription_msg_types: Set[Type[Command]] = {
+        commands.GetBeaconBlocks,
+    }
+
+    def __init__(self,
+                 db: BaseBeaconChainDB,
+                 peer_pool: BCCPeerPool,
+                 token: CancelToken = None) -> None:
+        super().__init__(peer_pool, token)
+        self.db = db
+
+    async def _handle_msg(self, base_peer: BasePeer, cmd: Command,
+                          msg: protocol._DecodedMsgType) -> None:
+        peer = cast(BCCPeer, base_peer)
+
+        if isinstance(cmd, commands.GetBeaconBlocks):
+            await self._handle_get_beacon_blocks(peer, cast(Dict[str, Any], msg))
+        else:
+            raise Exception("Invariant: Only subscribed to GetBeaconBlocks")
+
+    async def _handle_get_beacon_blocks(self, peer: BCCPeer, msg: Dict[str, Any]) -> None:
+        if not peer.is_operational:
+            return
+
+        max_blocks = cast(int, msg["max_blocks"])
+        block_slot_or_hash = msg["block_slot_or_hash"]
+
+        if isinstance(block_slot_or_hash, int):
+            get_start_block = functools.partial(
+                self.db.get_canonical_block_by_slot,
+                cast(int, block_slot_or_hash),
+            )
+        elif isinstance(block_slot_or_hash, bytes):
+            get_start_block = functools.partial(
+                self.db.get_block_by_hash,
+                cast(Hash32, block_slot_or_hash),
```

Comment:
In general `Hash32(block_slot_or_hash)` is to be preferred as this checks the type and e.g. `Hash32("foo")` will be rejected by mypy. In contrast nonsense such as `cast(Hash32, "foo")` is allowed as a cast will accept anything you feed it.

However, in this specific case, unfortunately the type hints don't have any affect at all. That is because `functool.partial` erases any type hints, everything becomes `Any` and we are left without any type safety. E.g. you can try replacing `cast(Hash32, block_slot_or_hash)` with `"foo"` and it will go through even though `get_block_by_hash` should not accept a string.

My preference would be to rewrite this part in a less dynamic way (reject wrong types, then execute one or the other)